apps/system/js/advanced_telemetry.js

```diff
@@ -484,29 +486,48 @@
   }
 
   AdvancedTelemetryPing.prototype.send = function(xhrAttrs) {
+    return new Promise((resolve, reject) => {
+      this.data = JSON.stringify(this.packet);
+      LazyLoader.load(['/shared/js/gzip/gzip.js']).then(() => {
+        Gzip.compress(this.data).then((gzipData) => {
+          debug('Length of compressed data:', gzipData.length);
+          return this.post(xhrAttrs, gzipData);
+        }, function (msg) {
+          debug('Error compressing telemetry payload:', msg);
+        }).then(function () {
+          resolve();
+        });
+      });
+    });
+  };
+
+  AdvancedTelemetryPing.prototype.post = function post(xhrAttrs, gzipData) {
     var xhr = new XMLHttpRequest({ mozSystem: true, mozAnon: true });
+    return new Promise((resolve, reject) => {
 
-    xhr.open('POST', this.url);
+      xhr.open('POST', this.url);
 
-    if (xhrAttrs && xhrAttrs.timeout) {
-      xhr.timeout = xhrAttrs.timeout;
-    }
+      if (xhrAttrs) {
+        xhr.onload = xhrAttrs.onload;
+        xhr.onerror = xhrAttrs.onerror;
+        xhr.onabort = xhrAttrs.onabort;
+        xhr.ontimeout = xhrAttrs.ontimeout;
 
-    xhr.setRequestHeader('Content-type', 'application/json');
-    xhr.responseType = 'text';
+        if (xhrAttrs.timeout) {
+          xhr.timeout = xhrAttrs.timeout;
+        }
+      }
+      xhr.setRequestHeader('Content-type', 'application/json');
 
-    var data = JSON.stringify(this.packet);
-    xhr.send(data);
-    //TODO:  GZIP COMPRESS.
+      try {
+        xhr.setRequestHeader('Transfer-Encoding', 'gzip');
+      } catch (e) {}
```

Comment:
It only throws during the unit test. With that said, I would like to understand better why it throws during the unit test because I would very much like to remove this try/catch.